File: src/algorithms/dynamic-programming/string-dp/word-break-memoization/word-break-memoization.test.ts
import { describe, it, expect } from "vitest";
import { wordBreakMemoization } from "./sources/word-break-memoization.ts?fn";

describe("wordBreakMemoization", () => {
  it("returns true for the default input 'leetcode' with ['leet', 'code']", () => {
    expect(wordBreakMemoization("leetcode", ["leet", "code"])).toBe(true);
  });

  it("returns false when no segmentation is possible", () => {
    expect(wordBreakMemoization("catsandog", ["cats", "dog", "sand", "and", "cat"])).toBe(false);
  });

  it("returns true for an empty string", () => {
    expect(wordBreakMemoization("", ["leet", "code"])).toBe(true);
  });

  it("returns true when text exactly matches one dictionary word", () => {
    expect(wordBreakMemoization("leet", ["leet", "code"])).toBe(true);
  });

  it("returns false when text has no matching words in dictionary", () => {
    expect(wordBreakMemoization("abcd", ["leet", "code"])).toBe(false);
  });

  it("returns true for 'applepenapple' with ['apple', 'pen']", () => {
    expect(wordBreakMemoization("applepenapple", ["apple", "pen"])).toBe(true);
  });

  it("returns true for 'catsanddog' with ['cat', 'cats', 'and', 'sand', 'dog']", () => {
    expect(wordBreakMemoization("catsanddog", ["cat", "cats", "and", "sand", "dog"])).toBe(true);
  });

  it("returns false for 'aaaaab' with only 'a' prefixes in dictionary", () => {
    expect(wordBreakMemoization("aaaaab", ["a", "aa", "aaa", "aaaa"])).toBe(false);
  });

  it("returns true when multiple segmentations are possible", () => {
    expect(wordBreakMemoization("pineapple", ["pine", "apple", "pineapple"])).toBe(true);
  });

  it("returns false when text is one character not in dictionary", () => {
    expect(wordBreakMemoization("z", ["a", "b", "c"])).toBe(false);
  });

  it("handles dictionary with single-character words", () => {
    expect(wordBreakMemoization("abc", ["a", "b", "c"])).toBe(true);
  });

  it("matches expected results for several standard inputs", () => {
    const cases: Array<[string, string[], boolean]> = [
      ["leetcode", ["leet", "code"], true],
      ["applepenapple", ["apple", "pen"], true],
      ["catsandog", ["cats", "dog", "sand", "and", "cat"], false],
      ["", ["word"], true],
    ];
    for (const [text, dictionary, expected] of cases) {
      expect(wordBreakMemoization(text, dictionary)).toBe(expected);
    }
  });
});

File: src/algorithms/dynamic-programming/string-dp/word-break-memoization/__tests__/word_break_memoization_test.py
import importlib
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "sources"))
mod = importlib.import_module("word-break-memoization")
word_break_memoization = mod.word_break_memoization

assert word_break_memoization("leetcode", ["leet", "code"]) == True, "leetcode should return True"
assert word_break_memoization("catsandog", ["cats", "dog", "sand", "and", "cat"]) == False, "catsandog should return False"
assert word_break_memoization("", ["leet", "code"]) == True, "empty string should return True"
assert word_break_memoization("leet", ["leet", "code"]) == True, "exact match should return True"
assert word_break_memoization("abcd", ["leet", "code"]) == False, "no match should return False"
assert word_break_memoization("applepenapple", ["apple", "pen"]) == True, "applepenapple should return True"
assert word_break_memoization("catsanddog", ["cat", "cats", "and", "sand", "dog"]) == True, "catsanddog should return True"
assert word_break_memoization("aaaaab", ["a", "aa", "aaa", "aaaa"]) == False, "aaaaab should return False"
assert word_break_memoization("pineapple", ["pine", "apple", "pineapple"]) == True, "pineapple should return True"
assert word_break_memoization("abc", ["a", "b", "c"]) == True, "single chars should return True"

if __name__ == "__main__":
    print("All tests passed!")
